lib/MusicBrainz/Server/Controller/Role/CommonsImage.pm

```diff
@@ -21,24 +25,53 @@ sub commons_image : Chained('load') PathPart('commons-image')
     $c->stash->{template} = 'components/commons-image.tt';
 }
 
-sub _get_commons_image
-{
+sub _get_commons_image {
     my ($self, $c, $cache_only) = @_;
 
     my $entity = $c->stash->{entity};
+    my $title;
+
+    # Check if there's a Wikimedia Commons relationship
     my ($commons_link) = map {
             $_->target;
         } grep {
             $_->target->isa('MusicBrainz::Server::Entity::URL::Commons')
         } @{ $entity->relationships_by_link_type_names('image') };
-
     if ($commons_link) {
-        my $page_name = $commons_link->page_name;
-        my $commons_image = $c->model('CommonsImage')->get_commons_image($page_name, cache_only => $cache_only);
-        if ($commons_image) {
-            $c->stash->{image} = $commons_image;
-        } else {
-            $c->stash->{image_url} = $c->uri_for_action($self->action_for('commons_image'), [ $entity->gid ]);
+        $title = $commons_link->page_name;
+    }
+
+    # and if not, check Wikidata entity
+    $title = _get_wikidata_image($c) if !$title;
+
+    my $commons_image = $c->model('CommonsImage')->get_commons_image($title, cache_only => $cache_only);
+    if ($commons_image) {
+        $c->stash->{image} = $commons_image;
+    } else {
+        $c->stash->{image_url} = $c->uri_for_action($self->action_for('commons_image'), [ $entity->gid ]);
+    }
+}
+
+sub _get_wikidata_image {
+    my ($c) = @_;
+    my $entity = $c->stash->{entity};
+
+    my ($wikidata_link) = map {
+            $_->target;
+        } grep {
+            $_->target->isa('MusicBrainz::Server::Entity::URL::Wikidata')
+        } @{ $entity->relationships_by_link_type_names('wikidata') };
+
+    if ($wikidata_link) {
+        my $properties = $c->model('WikidataProperties')->get_wikidata_properties($wikidata_link->pretty_name);
+        if ($entity->isa('MusicBrainz::Server::Entity::Area') &&
```

Comment:
Well, dunno if any of the other entity types would even have P242, but seems okay to only show those for areas.